test/metabase/test/util.clj

```diff
@@ -282,21 +282,24 @@
   "Temporarily set the value of the Setting named by keyword `setting-k` to `value` and execute `f`, then re-establish
   the original value. This works much the same way as `binding`.
 
-   Prefer the macro `with-temporary-setting-values` over using this function directly."
+  Prefer the macro `with-temporary-setting-values` over using this function directly."
   {:style/indent 2}
   [setting-k value f]
   ;; plugins have to be initialized because changing `report-timezone` will call driver methods
   (initialize/initialize-if-needed! :db :plugins)
-  (let [setting        (#'setting/resolve-setting setting-k)
-        original-value (when (or (#'setting/db-or-cache-value setting)
-                                 (#'setting/env-var-value setting))
-                         (setting/get setting-k))]
+  (let [setting                    (#'setting/resolve-setting setting-k)
+        env-var-value              (#'setting/env-var-value setting)
+        original-db-or-cache-value (#'setting/db-or-cache-value setting)]
     (try
       (setting/set! setting-k value)
       (testing (colorize/blue (format "\nSetting %s = %s\n" (keyword setting-k) (pr-str value)))
-        (f))
+        (if env-var-value
+          (testing (colorize/red (format "NOTE: %s is set to %s by an env var, which takes precedence\n"
+                                         (keyword setting-k) (pr-str env-var-value)))
+            (f))
+          (f)))
       (finally
-        (setting/set! setting-k original-value)))))
+        (setting/set! setting-k original-db-or-cache-value)))))
```

Comment:
Calling this out as a somewhat non-obvious change. Since the env var value now takes precedence when reading a setting, `do-with-temporary-setting-value` would pick it up and then write it back to the DB in the `finally` step. Instead we should just read the DB value directly and write it back at the end of the function.

I've added additional logging if a setting has an env var value set that overrides the temporary setting value, to hopefully avoid tricky testing bugs in the future.